fdbclient/FileBackupAgent.actor.cpp

```diff
@@ -6091,16 +6091,16 @@ class FileBackupAgentImpl {
 		} else {
 			TraceEvent("AS_StartRestore").log();
 			state Standalone<VectorRef<KeyRangeRef>> restoreRange;
-			bool containsSystemKeys = false;
+			state Standalone<VectorRef<KeyRangeRef>> systemRestoreRange;
 			bool encryptionEnabled = cx->clientInfo->get().isEncryptionEnabled;
 			for (auto r : ranges) {
 				if (!encryptionEnabled || !r.intersects(getSystemBackupRanges())) {
 					restoreRange.push_back_deep(restoreRange.arena(), r);
 				} else {
-					containsSystemKeys = true;
+					systemRestoreRange.push_back_deep(systemRestoreRange.arena(), r);
```

Comment:
This might be good enough in practice, though if we wanted to make it more bulletproof you could detect whether the intersecting range includes both system and non-system keys and then split it. Maybe something like:

```
if (!encryptionEnabled) {
    // add entire range to restore ranges
} else {
    KeyRangeRef normalKeyRange = r & normalKeys;
    KeyRangeRef systemKeyRange = r & systemKeys;
    if (!normalKeyRange.empty()) {
        // add to normal ranges
    }
    if (!systemKeyRange.empty()) {
        // add to system ranges
    }
}
```